File: node/producer.js
/**
 * Push tasks into an EZRA queue.
 *
 * Each task has:
 *   id        - sequential number
 *   duration  - simulated work time in seconds (0.1-10, log-uniform)
 *   will_fail - if true, the worker throws, triggering the retry/dead-letter path
 *
 * About 3% of tasks (id % 33 === 0) are marked will_fail so dead-letter
 * behavior shows up deterministically.
 *
 * Environment:
 *   EZRA_HOST    default: localhost
 *   EZRA_PORT    default: 42002
 *   EZRA_QUEUE   default: tasks
 *   TASK_COUNT   default: 50
 */

import { createClient } from 'redis'

const host  = process.env.EZRA_HOST  || 'localhost'
const port  = parseInt(process.env.EZRA_PORT  || '42002', 10)
const queue = process.env.EZRA_QUEUE || 'tasks'
const count = parseInt(process.env.TASK_COUNT || '50', 10)

// Deterministic log-uniform duration in [0.1, 10] seconds.
// Same seed always produces the same duration, so runs are reproducible.
function taskDuration(seed) {
  let h = (seed * 2654435761) >>> 0
  h = (h ^ (h >>> 16)) >>> 0
  const x = h / 4294967296
  return Math.round(Math.exp(Math.log(0.1) + x * Math.log(100)) * 100) / 100
}

const client = createClient({ socket: { host, port } })
await client.connect()

let willFailTotal = 0
console.log(`[producer] pushing ${count} tasks to ${host}:${port}/${queue}`)

for (let i = 1; i <= count; i++) {
  const will_fail = (i % 33 === 0)
  if (will_fail) willFailTotal++

  const payload = JSON.stringify({ id: i, duration: taskDuration(i), will_fail })
  await client.xAdd(queue, '*', { payload })

  if (i % 10 === 0) console.log(`[producer] ${i}/${count}`)
}

console.log(
  `[producer] done  total=${count}  ` +
  `will_fail=${willFailTotal} (-> dead-letter after 3 attempts)`
)

await client.disconnect()

File: python/producer.py
"""
Push tasks into an EZRA queue.

Each task has:
  id        - sequential number
  duration  - simulated work time in seconds (0.1-10, log-uniform)
  will_fail - if true, the worker raises, triggering the retry/dead-letter path

About 3% of tasks (id % 33 == 0) are marked will_fail so dead-letter behavior
shows up deterministically without having to wait for random failures.

Environment:
  EZRA_HOST    default: localhost
  EZRA_PORT    default: 42002
  EZRA_QUEUE   default: tasks
  TASK_COUNT   default: 50
"""

import json
import math
import os
import random

import redis


QUEUE = os.getenv("EZRA_QUEUE", "tasks")


def task_duration(seed: int) -> float:
    rng = random.Random(seed)
    return round(math.exp(rng.uniform(math.log(0.1), math.log(10.0))), 2)


def main() -> None:
    host = os.getenv("EZRA_HOST", "localhost")
    port = int(os.getenv("EZRA_PORT", "42002"))
    count = int(os.getenv("TASK_COUNT", "50"))

    client = redis.Redis(host=host, port=port, decode_responses=True, protocol=3)

    will_fail_total = 0
    print(f"[producer] pushing {count} tasks to {host}:{port}/{QUEUE}")

    for i in range(1, count + 1):
        will_fail = (i % 33 == 0)
        if will_fail:
            will_fail_total += 1

        client.xadd(QUEUE, {"payload": json.dumps({
            "id":        i,
            "duration":  task_duration(i),
            "will_fail": will_fail,
        })})

        if i % 10 == 0:
            print(f"[producer] {i}/{count}")

    print(
        f"[producer] done  total={count}  "
        f"will_fail={will_fail_total} (-> dead-letter after {3} attempts)"
    )


if __name__ == "__main__":
    main()
